parse serviceName in VerticaReplicator

fenic-fawkes opened the pull request:
allow users to provide a service name to use for source and target databases. if unspecified we use the first service assigned to the database with a cluster ip, if supplied we validate that it's assigned to the appropriate main/sandbox cluster

Commit:
VCluster API Refresh (#741)

This pulls in the latest vclusterops library changes. Many of the
changes that we are pulling in were changes related to improving the
vcluster CLI.

Commit:
Merge branch 'main' into vnext

jizhuoyu committed:
VerticaReplicator Skeleton API and Controller (#740)

Commit:
Add replicated upgradePolicy (#743)

Commit:
Merge branch 'main' into vnext

Commit:
VCluster API Refresh

This pulls in the new signature for VStartDatabase. A new parameter was
returned, which we can safely ignore for the operators usage.

Commit:
Assign subclusters to replica groups for replicated upgrade (#748)

During a replicated upgrade, we must categorize the subclusters into one
of two replica groups. Initially, all subclusters participating in the
upgrade will be placed in replica group A. Subsequently, new subclusters
will be generated throughout the process and assigned to the second
group, named replica group B. This change manages the assignment of
subclusters at the beginning of the upgrade process. To facilitate this
change, new status fields will be introduced to the VerticaDB, enabling
the tracking of this state across reconcile iterations.

roypaulin committed:
Add new sandbox controller (#749)

This adds a new controller for sandbox. It is going to watch a ConfigMap
that has labels on it to indicate it contains state for a sandbox. That
ConfigMap also contains the vdb name whose subclusters are involved in
sandbox/unsanbox. For now it simply reads the VerticaDB found in the
configMap and do nothing else.
Some changes are also made to the VerticaDB API to add new fields
related to sanbox

jizhuoyu committed:
Remove labels in `vrep` sample config (#753)

This PR eliminates labels from the `VerticaReplicator` sample config.
The intention is to enhance the user experience on OpenShift when
generating this CR from the webUI. OpenShift utilizes this sample to
prepopulate fields in the webUI interface.

Commit:
Add webhook rules for replicated upgrade (#755)

This adds a few webhook rules for replicated upgrade. It's probably not
a complete list, but we can always add more later when we find rules to
check. This commit adds the following:
- ensures the subclusters involved in the upgrade are stable. They
cannot be removed and the sizes can't change.
- We do allow new subclusters to be added to one of the replica groups.
But they must be secondary subclusters. No new primaries.
- ensures that subclusters listed in the
`.status.updateState.replicaGroups` aren't repeated.

Commit:
Merge branch 'main' into vnext

Commit:
Merge branch 'main' into vnext

roypaulin committed:
Collect sandbox name in podfacts (#762)

We now collect the sandbox name in podfacts. We updated the vsql query
to return the extra state. We built an interface to get the information
as we intend to use a REST endpoint in the future but still need support
via vsql for older releases.

Commit:
Add new leg for 24.3.0 tests (#763)

This adds a new leg, e2e-leg-9, to the CI. This is going to be used to
test changes done in 24.3.0. One new change with this is that it will
use a vertica license so that we can test scaling past 3 nodes. This
will be required in order to test out the new replicated upgrade work.

jizhuoyu committed:
`vrep` reconciler enforces minimum version of source and target `vdb`s (#764)

`vrep` reconciler enforces the minimum source and target db versions to
be at least 24.3.0

---------

Co-authored-by: Roy Paulin <rnguetsopken@opentext.com>
Co-authored-by: spilchen <mspilchen@opentext.com>

Commit:
Replicated upgrade: create secondaries based on primaries (#765)

cchen-vertica committed:
Add webhook rules for sandboxes (#767)

This PR adds below webhooks rules for sandboxes in CRD:
- cannot scale (up or down) any subcluster that is in a sandbox
- cannot remove a subcluster that is sandboxed. It must be unsandboxed
first.
- cannot have multiple sandboxes with the same name
- cannot have the image of a sandbox be different than the main cluster
before the sandbox has been setup
- cannot be used on versions older than 24.3.0
- cannot be used before the database has been initialized
- cannot have duplicate subclusters defined in a sandbox
- cannot have a subcluster defined in multiple sandboxes
- cannot have a non-existing subcluster defined in a sandbox

We could add more rules later for sandboxes when we have needs.

roypaulin committed:
Add restart reconciler to sandbox controller (#766)

we want to use restart reconciler in both the VerticaDB controller and
the sandbox controller. This makes some changes to the restart
reconciler in order to achieve that. The sandbox name (empty string if
VerticaDB controller) is passed to the reconciler and it uses it to
target only pods belonging to that sandbox( or to target only pods
belonging to the main cluster if sandbox name is empty)

Commit:
Add stubs for replicated upgrade (#768)

This change adds stubs to the replicated upgrade reconciler. It attempts
to have the functions present that are needed to do the upgrade. Due to
dependencies, such as sandboxing and replication, comprehensive testing
is currently limited to unit tests. While this logic might evolve during
integration, it gives a framework for how the upgrade should function.

Additionally, I identified new requirements for state management. Two
new annotations have been introduced to the VerticaDB:
- vertica.com/replicated-upgrade-sandbox: to record the name of the
sandbox created for the upgrade
- vertica.com/replicated-upgrade-replicator-name: to track the name of
the VerticaReplicator CR utilized for replicating data to the sandbox.

cchen-vertica committed:
Add a new webhook rule for sandboxes (#770)

This PR adds a new webhook rule for sandboxes in CRD: cannot have a
primary subcluster defined in a sandbox.

cchen-vertica committed:
Use VClusterOps API to get node details in podfacts (#769)

This adds a new fetcher in podfacts to fetch node details from the
running database inside the pod. This new fetcher will call VClusterOps
API which will send an HTTP request to the database. The new fetcher
will be enabled when VclusterOps annotation is set, and Vertica server
version is not older than v24.3.0. The new fetcher should be faster and
more reliable than the old fetcher which will execute vsql within the
pod.

Commit:
Redirect clients to replica group B during replicated upgrade (#771)

In the replicated upgrade process, we must pause and redirect client
connections from subclusters in replica group A to those in replica
group B. This is the initial stage of the change. Currently,
pause/redirect semantics are not supported, as they require new server
modifications that have not yet been implemented. Therefore, we will
perform an old-style drain to pause the process and maintain service
labels correctly to point to subclusters in replica group B for
redirection.

---------

Co-authored-by: Roy Paulin <rnguetsopken@opentext.com>

roypaulin committed:
Update upgrade reconciler to handle subclusters in sandbox (#772)

This change will modify the upgrade reconcilers, both offline and
online, to function in either the main cluster or a sandbox. Our
immediate plan is to use the offline upgrade reconciler within the
sandbox controller, although this will be done in a follow-on task.

Commit:
Merge branch 'main' into vnext

jizhuoyu committed:
Add replication reconciler for VerticaReplicator (#773)

This pull request modifies the Vertica replicator controller to invoke
the replication command synchronously. Additionally, it introduces new
status conditions within VerticaReplicator to monitor the controller's
various states.

---------

Co-authored-by: spilchen <mspilchen@opentext.com>

Commit:
Add event when different upgrade policy is picked (#776)

If we had selected an upgrade method but ended up failing back to a
different method because something was incompatible, we are going to log
an event. For instance, if we requested something other than offline
upgrade, but we end up in the offline upgrade code path we will have an
event message like this:
```
 Normal   IncompatibleUpgradeRequested  7m20s                  verticadb-operator  Requested upgrade is incompatible with the Vertica deployment. Falling back to offline upgrade.
```

cchen-vertica committed:
Add Sandbox Reconciler to VerticaDB Controller (#775)

This PR adds sandbox reconciler to VerticaDB controller. The reconciler
will add subclusters to sandboxes in the database. The change of sandbox
fields in CRD will trigger sandbox reconciler, then the reconciler will
call vclusterOps to add subclusters to sandboxes.

Commit:
Add status messages for replicated upgrade (#778)

This will add a status message to the `.status.upgradeStatus` field in
the VerticaDB as the operator goes through the various stages of
replicated upgrade. It reuses the same design for the other upgrade
methods in that we only advance the status message forward. This means
that we need to reconcile an iteration, we won't report the first status
message again.

github-actions[bot] committed:
Change operator version to 2.2.0 (#784)

Automated changes by pre-release.yml GitHub action

Signed-off-by: GitHub <noreply@github.com>
Signed-off-by: Matt Spilchen <mspilchen@opentext.com>
Co-authored-by: spilchen <spilchen@users.noreply.github.com>

Commit:
Merge branch 'main' into vnext

jizhuoyu committed:
e2e tests for `vrep` (#779)

testing for the following cases (no custom user name and custom password
secret in `vrep` specified, the source vdb itself uses a custom k8s
secret for superuser password):
- replicate from main cluster of a vdb to main cluster of another vdb
- replicate from main cluster of a vdb to a sandbox cluster of another
vdb

Commit:
Decouple subcluster name from its statefulset name (#780)

This adds a new subcluster annotation to control the statefulset name.
The new annotation is: `vertica.com/statefulset-name-override`. This is
needed for replicated upgrade when the sandbox gets promoted to the main
cluster. We will want to rename the sandbox subcluster to match the
original names. However, the statefulset name always included the
subcluster name. This new annotation will allow us to rename the
subcluster while keeping the statefulset name constant. This ensures we
can do a rename without needing to rebuild the statefulset.

chinhtranvan committed:
Give the vdb verification exactly once semantics for vrep (#782)

To follow up on #781, we will verify the presence of QueryReady instead
of checking if QueryReady is true. This ensures the verification of
exactly-once semantics for vrep.

roypaulin committed:
Add offline upgrade reconciler to sandbox controller (#785)

This will allow us to upgrade the subclusters in a sandbox. There is a
new image field in the VerticaDB for the sandbox. When that is changed,
it will cause the sandbox controller to drive the upgrade. The vdb
controller will check if a sandbox needs an upgrade and will wake up the
sandbox controller by setting the vdb resource version in the configmap
annotations.
You can follow a sandbox upgrade progress through the upgradestate field
in the sandbox status.

chinhtranvan committed:
Vcluster API Refresh (#792)

I have a GitHub issue , then I moved #789 to this PR
This pulls in the latest vclusterops library changes. Many of the
changes that we are incorporating relate to improvements made to the
vcluster CLI:

- Replaced pointer types with value types for all database operations.
- Remove oldIPv6 and update the tests
- Add eon option for remove_sc_vc.go to integrate the changes in
remove_subcluster CLI

Commit:
Use static labels in pods (#791)

During a replicated upgrade we will promote a sandbox to the main
cluster, which cause it to drop its sandbox state, and we need to rename
subclusters to match the original names. Prior to this change, the pods
deployed by the operator had labels for the sandbox and subcluster name.
And the labels were set in the statefulset pod template. Now, in order
to change those, the k8s controller will actually do a rollout. This
means it needs to restart each of the pods. But we require sandbox
promotion and subcluster rename to happen without a pod restart. So, the
change here to use different labeling to ensure we don't need to restart
the pods. In some cases, if we have a pod we have to refer back to the
statefulset in order to figure certain metadata.

Since we are changing the statefulset selector values, the statefulset
from old operator need to be recreated. We had done this in the past, so
I repurposed the upgradeoperator120_reconciler.go for this purpose.

A summary of all of the labels that are set is as follows.

- All objects will have these labels set:
```
app.kubernetes.io/instance: <CR name>
app.kubernetes.io/managed-by: verticadb-operator
app.kubernetes.io/component: database
app.kubernetes.io/name: vertica
app.kubernetes.io/version: 2.2.0
vertica.com/database: <vertica db name>
```
- StatefulSets and Service objects will have the following labels:
```
vertica.com/subcluster-name: <subcluster-name>
vertica.com/subcluster-type: <primary|secondary>
vertica.com/sandbox: <sandbox name if applicable>
```
- Further Service objects will have this additional label:
```
vertica.com/svc-type: <external|headless>
```
- Pods will have the following labels:
```
vertica.com/subcluster-svc: <name of service object>
vertica.com/client-routing: true # Only set if not doing a drain or pending removal
vertica.com/subcluster-selector-name: <full name of the pod's statefulset>
```
- ConfigMaps used by the sandbox controller will have this label:
```
vertica.com/watched-by-sandbox-controller: true
```
- When the statefulset sets up the pod selector, it will use these
labels:
```
app.kurbernetes.io/instance
vertica.com/subcluster-selector-name
```
- When the service objects sets up its pod selector, it will use these
labels:
```
app.kurbernetes.io/instance
vertica.com/subcluster-selector-name
vertica.com/client-routing
```

chinhtranvan committed:
fix tests (#794)

This is a quick fix. When I merge my branch
(https://github.com/vertica/vertica-kubernetes/pull/792) into vnext, I
don't pull the latest changes from Roy's PR
(https://github.com/vertica/vertica-kubernetes/pull/785). This PR will
resolve the e2e tests issue

cchen-vertica committed:
Create sandbox config maps in sandbox reconciler (#790)

This will create/update a sandbox config map whenever we create a new
sandbox in sandbox reconciler. The config map is watched by sandbox
controller for restarting/upgrading the nodes in a sandbox. The config
map will be deleted if vdb is deleted or the sandbox is unsandboxed.

jizhuoyu committed:
add test for `vrep` custom username / password and tls config (#786)

test the following scenarios:
- replicate from main cluster of source db with default username
(`dbadmin`) to a sandbox cluster of target db with custom username and
no password
- replicate from main cluster of source db with default username
(`dbadmin`) to a sandbox cluster of target db with custom username and
password
- replicate from main cluster of source dbwith default username
(`dbadmin`) to a sandbox cluster of target db with default username
(`dbadmin`) and source TLS config

chinhtranvan committed:
Update sandbox label (#798)

A nondeterministic failure occurs when a label is updated based on the
sandbox spec. We need to ensure that the label is updated only after the
status is set by the sandbox subcluster

chinhtranvan committed:
Add sandbox option in vrep (#795)

This PR incorporates the latest changes from the vCluster API to update
the sandbox option in vrep

Commit:
Merge branch 'main' into vnext

Commit:
Set VerticaDB name in ConfigMap (#802)

The ConfigMap should have the name of the VerticaDB, not the database
name. The sandbox controller uses this to figure out the correct
VerticaDB to fetch.

Commit:
Add support for multiple subclusters in a sandbox (#799)

This adds support for multiple subclusters in a single sandbox. In order
to denote which subclusters are primary in the sandbox, a new subcluster
type is added called 'sandboxprimary'. New webhook rules have been added
to account for the new subcluster type.

The new type will get added to the first subcluster in a sandbox. We now
add subclusters to a sandbox in the order they are defined in the vdb
spec. This ensures that we pick a deterministic subcluster as the first
one in a sandbox.

When adding a subcluster to an existing sandbox, we need to pass in a
host from the main cluster and the sandbox too. This reason this is
needed is that we need to replicate the new sandbox state in both the
main cluster and in the sandbox. So, the vclusterOps API will update
both clusters.

Commit:
Use ForceRemoval option with revive (#804)

This was mistakenly removed in the last VCluster API refresh (#741).
Without this, if there is an error during revive, then it will complain
that local catalog/data/depot dirs are not empty.

roypaulin committed:
Add status reconciler to sandbox controller (#793)

The status reconciler in vdb controller was removing a subcluster's
status when the subcluster is in a sandbox. Now it will keep unchanged
the status of subclusters that do not belong to the passed in sandbox.
To do that, it finds all the subclusters regardless of their cluster.

jizhuoyu committed:
Add test step for `verify-vrep` to verify replication from empty db (#806)

chinhtranvan committed:
Vcluster-API-Refresh (#801)

This incorporates the latest changes from the vclusterops library. Many
of these changes affect e2e tests. We are pulling the latest updates to
incorporate these changes.
* Limit version check for down nodes in list_allnode to only CLI
* Revert pointer adapter pools because it causes multiple operator
threads in the operator
* Put vdb.hosts in prepare in HTTPSUpdateNodeStateOp since it will be
empty, and the operator can't find hosts to execute this operation
* Add username to HTTPS operation for re_ip, since the operator can't
verify username-password authentication for running httpscheckrunningDB.

Commit:
Call sandbox reconciler during replicated upgrade (#807)

This builds out the replicated upgrade logic more. With the sandbox
controller and the replicator controller both implemented, the upgrade
logic can now depend on them. A few changes were made while verifying
the new controllers. We can now get to the part where we are waiting for
promotion of the sandbox to the main cluster. All traffic is routed to
the sandbox, which have been upgraded to the new server version.

roypaulin committed:
Decouple sandbox image from vdb.spec image (#812)

When setting a sandbox, if the image is omitted, the operator will
internally assume spec.image as the sandbox's image. The issue is that
changing spec.image could trigger sandbox upgrade too. Now if the image
is omitted, there is a new reconciler that will explicitly set the
sandbox image in vdb to spec.image value.

---------

Signed-off-by: roypaulin <roy.paulin@microfocus.com>

cchen-vertica committed:
Add unsandbox reconciler to vdb and sandbox controllers (#803)

This PR added two unsandbox reconcilers: one in vdb controller and the
other in sandbox controller. The one in vdb controller will update
sandbox config map for triggering sandbox controller. The one in sandbox
controller will call vclusterOps to unsandbox the subclusters, update
vdb sandbox status, and delete the config map. After that, the restart
reconciler in vdb controller should restart the nodes that have been
unsandboxed.

roypaulin committed:
Use subcluster status to update podfacts (#814)

After unsandbox, the catalog dir is deleted. The issue we check its
existence to set dbExists in the pod facts. With this change, dbExists
is true if the catalog dir exists or if addedToDB field in the
subcluster status detail for that specific pod is true.

Commit:
Fix for many subclusters sandboxed at the same time (#815)

This update addresses an issue that occurs when multiple subclusters are
added to a new sandbox during the same reconciliation iteration. The
problem arose after sandboxing the first subcluster; the second
subcluster failed to sandbox.

The failure was due to the need to provide both a host from the main
cluster and a host from the sandbox to the vcluster API. To obtain the
sandbox host, the system relies on the `vdb.status.sandbox` being
updated. Since this update wasn't happening, the process of finding the
sandbox host by checking pod facts didn't locate any sandboxed pods.

The solution is to update `vdb.status.sandboxes` as each subcluster is
sandboxed. This update allows subsequent sandboxes to find the correct
pod facts, ensuring the process completes successfully for all
subclusters.

---------

Co-authored-by: Roy Paulin <rnguetsopken@opentext.com>

cchen-vertica committed:
Applied new option --hosts-in-sandbox of vclusterOps start_db on the operator (#813)

When restarting a pod in a database with a sandbox, it could fail when
the IP of the pod changes. This is a bug in old vclusterOps code. We
were trying to read node IPs from the catalog, but sandbox node IP
change cannot be reflected in the main cluster catalog, and vice versa.
This PR applied the new vclusterOps version and passed a new option to
start_db to fix this issue. With this option, we will read main cluster
node IPs from main cluster catalog and read sandbox node IPs from
sandbox catalog.

cchen-vertica committed:
Fixed polling timeout issue of start_db and start_node (#818)

In k8s, the node state polling timeout is controlled by the annotation
"vertica.com/restart-timeout". If it's not set, the polling timeout is
set to 0 which will fail vclusterOps start_db and start_node. Then we
will have many cluster/node restart failure events. This PR simply used
vclusterOps default polling timeout if the annotation
"vertica.com/restart-timeout" is not set.

roypaulin committed:
Add new test for sandbox upgrade (#816)

This adds tests for sandbox upgrade:
- `sandbox-upgrade-happy-path`: runs a successful sandbox upgrade
- `sandbox-upgrade-bad-image`: starts an upgrade with a bad image, abort
it and get it back to a good state

---------

Signed-off-by: roypaulin <rnguetsopken@opentext.com>

jizhuoyu committed:
Add `SourceVerticaDB`, `TargetVerticaDB`, `State`, `Age` print columns to `vrep` (#819)

roypaulin committed:
Add e2e test for sandbox restart (#821)

This adds a test for restarting sandbox.

cchen-vertica committed:
Fixed bugs in sandbox and unsandbox (#820)

This PR applies the latest vclusterOps and passed more node information
to vclusterOps for fixing the bugs in sandbox and unsandbox. The bugs
can be reproduced when we kill a pod in a subcluster before we sandbox
it and kill a pod in a sandbox before we unsandbox it. The issue is
cause by the stale in the catalog. The vcluster re-ip only can modify
the nodes' IPs on either main cluster or the sandbox. The new
vclusterOps will do one more re-ip before sandboxing/unsandboxing.

fenic-fawkes committed:
print a more useful error message if sandbox is supplied but empty of nodes (#822)

fenic-fawkes committed:
parse serviceName in VerticaReplicator

roypaulin committed:
Fully mirror main cluster in replicated upgrade (#824)

We used to mirror only primary subclusters. Now we mirror secondary
subclusters as well.

cchen-vertica committed:
Added e2e test for verifying replication in upgrade (#823)

In the replicated upgrade, the operator should trigger the replication
reconciler after the sandbox is upgraded. The replication reconciler
will create a replication CR and copy the data from the main cluster to
the sandbox. This PR added an e2e test to verify that replication
process.

roypaulin commented on pkg/controllers/vrep/replication_reconciler.go:
```diff
@@ -238,33 +236,67 @@ func (r *ReplicationReconciler) checkSandboxExists() error {
+		}
 	} else {
-		r.TargetInfo.IP = upPodIP
+		// assume target must not be read-only, no subcluster constraints
```
If service name is empty you should pick the first primary service.

roypaulin commented on pkg/controllers/vrep/replication_reconciler.go:
```diff
@@ -238,33 +236,67 @@ func (r *ReplicationReconciler) checkSandboxExists() error {
 	return nil
 }
 
+func (r *ReplicationReconciler) getServiceIP(ctx context.Context, svcName, dBname string) (ip string, err error) {
```
A simpler solution would be to just check if the service name belongs to one of the subclusters. If that's the case you fetch it and get the ip. You don't even need to check the labels.

roypaulin commented on pkg/controllers/vrep/replication_reconciler_test.go:
We should also add a test case for that.

fenic-fawkes replied:
are there examples of tests i could base one off of?

roypaulin replied:
There is a e2e test `verify-vrep`. You probably need to update that one. @jizhuoyu can help.

fenic-fawkes replied:
ok. i mostly got tests to work. i'm having trouble getting the test to generate a service for the sandbox with a cluster ip, but i'll keep working on it

fenic-fawkes replied:
oh, i'm a goomba. the services are being created fine but the statefulset isn't.
i could use help creating a sandbox in the tests, it doesn't look like there's an example in that file

roypaulin commented:
We should also update e2e test. Let's discuss about this at the standup.

cchen-vertica committed:
Fixed intermittent failure in online-upgrade-kill-transient test (#825)

This PR applied new vclusterOps library to fix intermittent failure in
online-upgrade-kill-transient test. That intermittent failure is caused
by IP inconsistency between the transient node and the catalog. In the
new vclusterOps library, we won't check the version of the transient
node in start_node, and we will do re-ip before removing the transient
node.

jizhuoyu commented on pkg/controllers/vrep/replication_reconciler.go:
```diff
@@ -238,33 +239,67 @@ func (r *ReplicationReconciler) checkSandboxExists() error {
+}
+
 // choose the source host and target host
 // (first host where db is up in the specified cluster)
```
could we update the comment to reflect the new logic related to service name?

jizhuoyu commented on pkg/controllers/vrep/replication_reconciler.go:
```diff
@@ -238,33 +239,67 @@ func (r *ReplicationReconciler) checkSandboxExists() error {
 	return nil
 }
 
+func (r *ReplicationReconciler) getServiceHostname(ctx context.Context, vdb *vapi.VerticaDB, sandboxName, svcName string) (ip string, err error) {
```
could we update the comment for this function to explain the behavior, especially when `svcName` is empty string?

jizhuoyu commented:
> _ No description provided. _

Could we update the PR description please?

roypaulin commented on pkg/controllers/vrep/replication_reconciler.go:
```diff
@@ -238,33 +239,74 @@ func (r *ReplicationReconciler) checkSandboxExists() error {
+		return "", err
+	}
+
+	for i := range stss.Items {
```
An easier way would be:
- If svcName is non-empty: to iterate over vdb.spec.subclusters, get the servicename(`names.GenExtSvcName(vdb, sc)`). check if it is equal svcName, check if the sc sandboxName is equal to sandboxName.
- If svcName is empty: iterate over vdb.spec.subclusters, pick the 1st primary subcluster's service. main cluster primaries have `type = primary` while sandbox primaries have `type = sandboxPrimary`.

To know if a sc belongs to a sandbox you need to check the status. There are already functions like `vdb.GenSubclusterSandboxStatusMap()` or `vdb.GetSandboxStatus()` that can help.

fenic-fawkes commented:
needs one small vcluster change that's under review but then this should be good to merge

chinhtranvan committed:
Downgrade Vertica Image for Temporary Fix of e2e-leg-9-vcluster (#832)

This PR updates the Vertica image to `20240606` to avoid replication
errors from the server side. This is a temporary fix, and we will revert
this change after fixing the issue on the server side

chinhtranvan committed:
Add reconciler to promote the sandbox to the main cluster (#828)

This PR build a reconciler so that during the upgrade we can just
promote the sandbox after all of the data has been replicated to it.

---------

Co-authored-by: cchen-vertica <cai.chen@microfocus.com>

roypaulin commented on pkg/vadmin/replication_start_vc.go:
```diff
@@ -68,11 +68,11 @@ func (v *VClusterOps) genReplicateDBOptions(s *replicationstart.Parms, certs *HT
 	opts.IsEon = v.VDB.IsEON()
 
-	opts.IPv6 = net.IsIPv6(s.SourceIP)
+	opts.IPv6 = net.IsIPv6(s.SourceHostname)
```
This only works with ips. Maybe we should just use the service clusterIP

fenic-fawkes replied:
my issue is that if we use IPs then it could break tls. if the tls config being used for on the source side is validating hostnames it could validate the service name but not the ip address. in general we should prefer hostnames when tls might be involved. should i just remove this check?

roypaulin commented on pkg/controllers/vrep/replication_reconciler.go:
```diff
@@ -238,33 +237,65 @@ func (r *ReplicationReconciler) checkSandboxExists() error {
+			sc := &vdb.Spec.Subclusters[i]
+			scSvcName := names.GenExtSvcName(vdb, sc)
+			if inputName == scSvcName && sandboxName == vdb.GetSubclusterSandboxName(sc.Name) {
+				return fmt.Sprintf("%s.%s.svc.cluster.local", scSvcName.Name, scSvcName.Namespace), nil
```
For precaution try to retrieve the service.

fenic-fawkes replied:
you mean the way that i did previously?

cchen-vertica commented:
I've synced the vclusterOps to the Github repository. You can update your go.mod to refer to the latest commit in vcluster repo.

cchen-vertica committed:
Closed user sessions before db replication starts in db upgrade (#834)

This PR used vsql to close all active user sessions before database
replication starts in the replicated upgrade. This can speed up the
upgrade process. This is a temporary change. We will improve this by
pausing all user sessions soon.

roypaulin commented on pkg/controllers/vrep/replication_reconciler.go:
```diff
@@ -238,33 +237,92 @@ func (r *ReplicationReconciler) checkSandboxExists() error {
+}
+
+// used below to avoid a really messy if condition
+func (r *ReplicationReconciler) shouldPickService(ctx context.Context, vdb *vapi.VerticaDB, sc *vapi.Subcluster,
+	sandboxName, svcName string) (should bool, err error) {
+	scSvcName := types.NamespacedName{Namespace: vdb.Namespace, Name: sc.ServiceName}
+	if scSvcName.Name == "" {
+		scSvcName = names.GenExtSvcName(vdb, sc)
+	}
+	if svcName != "" {
+		return scSvcName.Name == svcName, r.validateServiceIP(ctx, scSvcName)
+	}
+
+	if vdb.GetSubclusterSandboxName(sc.Name) != sandboxName {
+		return false, nil
+	}
+
+	return sc.IsPrimary(), r.validateServiceIP(ctx, scSvcName)
+}
```
```suggestion
func (r *ReplicationReconciler) shouldPickService(ctx context.Context, vdb *vapi.VerticaDB, sc *vapi.Subcluster,
	sandboxName, svcName string) should bool {
	if vdb.GetSubclusterSandboxName(sc.Name) != sandboxName {
		return false
	}
	if svcName != "" {
		return sc.GetServiceName() == svcName
	}

	return sc.IsPrimary(), r.validateServiceIP(ctx, scSvcName)
}
```

fenic-fawkes replied:
the issue is that `sc.GetServiceName()` only returns the subcluster part of the service name if `sc.ServiceName` is empty. `k get service` will list service names as `<db-name>-<sc-name>`, so we should be comparing against the full service name, no? maybe this is just an artifact of the tests, but the `names.GenExtSvcName` call was required to get them to pass

fenic-fawkes replied:
discussed with roy, i understand the desired behavior now. we should be comparing against the non-vdb portion of the service name

roypaulin reviewed (commented):
Very close. Sorry if what I say is often confusing. Let me know If I misunderstood what you are trying to do.

fenic-fawkes commented on pkg/controllers/vrep/replication_reconciler.go:
```diff
@@ -238,33 +237,92 @@ func (r *ReplicationReconciler) checkSandboxExists() error {
+		return scSvcName.Name == svcName, r.validateServiceIP(ctx, scSvcName)
+	}
+
+	if vdb.GetSubclusterSandboxName(sc.Name) != sandboxName {
```
the reason i did it this way is so that specifying a service is enough to direct traffic to the sandbox. if you feel strongly that we shouldn't allow that behavior i'll reverse this

cchen-vertica committed:
Removed old main cluster and renamed new main cluster in replicated upgrade (#829)

This PR added the last two steps of replicated upgrade: 
1. remove the old main cluster. After we promote the sandbox, the old
main cluster is not needed anymore so we remove it.
2. rename the subclusters in new main cluster. After we promote the
sandbox, we rename all subclusters in the sandbox to match the ones in
original main cluster.

chinhtranvan committed:
support subcluster name "default_subcluster" in replicated upgrade (#838)

This PR changes underscore to hyphen for the service name in the
operator.

chinhtranvan committed:
Revert Latest Vertica Image for E2E Leg9 vCluster (#840)

With the change on the server side, we have already fixed the issue in
https://github.com/vertica/vertica-kubernetes/pull/832. We can revert to
the latest Vertica image to fix the replication issue

fenic-fawkes committed:
use updated vcluster (tmp)

cchen-vertica committed:
Applied the latest vclusterOps library (#842)

The latest vclusterOps improves the promote_sandbox API to clean
communal storage after sandbox promotion. This PR simply updates
vclusterOps library to clean communal storage in the online upgrade.

roypaulin committed:
Remove sandbox configmap after promotion (#843)

After promoting sandbox to main, we need to delete sandbox config map to
avoid any kind of conflicts as the sandbox does not exist anymore.

roypaulin commented:
Almost done! Resolve the conflicts and run the tests again.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vertica/vertica-kubernetes?pullRequest=826) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/vertica/vertica-kubernetes?pullRequest=826) before we can accept your contribution.<br/>**5** out of **6** committers have signed the CLA.<br/><br/>:white_check_mark: jizhuoyu<br/>:white_check_mark: chinhtranvan<br/>:white_check_mark: cchen-vertica<br/>:white_check_mark: fenic-fawkes<br/>:white_check_mark: roypaulin<br/>:x: spilchen<br/><hr/>**spilchen** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vertica/vertica-kubernetes?pullRequest=826) it.</sub>
